### DISCUSSION TITLE
Multiple extensions within one application

### DISCUSSION THREAD

**Discussion by dallashuggins** [directus/directus#8550] [Archive]
What is currently supported as far as using multiple extensions? I have an interface in my application, but I was hoping to add a display within the same app. 

I read the discussion [here](https://github.com/directus/directus/discussions/3108), and added `package.json` files to each of the individual extensions, and added pack as the type for `directus:extension` in the root `package.json`. I wasn't sure if this was the intended setup, I likely am just structuring incorrectly, so any guidance would be much appreciated. 

_The structure of my app and extensions:_
```
extensions/
  displays/
    pack-price/
      src/
        index.ts (etc.)
    index.js (generated)
    package.json
  interfaces/
    price-conversion/
      src/
        index.ts (etc.)
    index.js (generated)
    package.json
node_modules/
package.json
```

In the root `package.json` I have specified:
```
  "directus:extension": {
    "type": "pack",
    "host": "^9.0.0-rc.95"
  },
```

The root `package.json` also has the following `dependencies`:
```
    "@directus/extension-pack-price": "1.0.0",
    "@directus/extension-price-conversion": "1.0.0",
```

The display `package.json`:
```
{
  "name": "@directus/extension-pack-price",
  "version": "1.0.0",
  "private": true,
  "directus:extension": {
    "type": "display",
    "path": "index.js",
    "source": "src/index.ts",
    "host": "^9.0.0-rc.95",
    "hidden": false
  }
}
```

The interface `package.json`:
```
{
  "name": "@directus/extension-price-conversion",
  "version": "1.0.0",
  "private": true,
  "directus:extension": {
    "type": "interface",
    "path": "index.js",
    "source": "src/index.ts",
    "host": "^9.0.0-rc.95",
    "hidden": false
  }
}
```

Is what I'm trying to do possible? If so, what am I doing wrong?

I keep getting the following error whenever I try to build. 
```
[Error] Current directory is not a Directus extension.
```

Thanks for any help!

**Comment by rijkvanzanten** [comment]
Cc @nickrum

**Comment by joselcvarela** [comment]
Hello @dallashuggins 
So definitely @nickrum is the best person to help with this.

But since you are using `package.json` this could be a misconfiguration on your project structure.
From what I am understanding from your structure, you have `package.json` and `index.js` inside `interfaces` and `displays`. So I think you need to move those inside you extensions. For instance, to inside `pack-price` and `price-comparison`.

**Reply by dallashuggins** [reply]
Aw yes, I actually messed that up in the code above since those are in the `pack-price` and `price-conversion`, but good catch. Thanks!

**Comment by dimitrov-adrian** [comment]
@dallashuggins 
It is supported right now, and pack extension type is more like metapackage, it have some ugly parts like the manifest still need path and entrypoint, but it works fine.

I already have extension that use almost same approach, but uses endpoint and interface. You could check https://github.com/dimitrov-adrian/directus-extension-linkmeta Where I am using local packages as a dependencies, the cons in this way is that you need to hoist the dependencies (this is why dependencies are added to the main package) to the metapackage as npm need update after install because the dependencies of local dependencies are not resolved during installation.

**Reply by dallashuggins** [reply]
Awesome, thanks for the example and details! Very helpful.

**Comment by nickrum** [comment]
Npm extensions aren't officially supported yet, so there might be issues with what you are trying to do. The easiest way to use multiple extensions right now is to use local extensions (the ones you put into the `extensions/` folder).

You can't use the `@directus/extension-` prefix for npm extensions because the `@directus` scope is reserved for official directus packages. You have to use the `directus-extension-` prefix instead.

The extensions-sdk doesn't currently support building `pack` extensions as they only serve as a virtual wrapper around actual extensions and there is no need to build them, hence the error message.

Regarding your package structure, I would suggest going with a monorepo structure using npm workspaces. The `pack` extension and all of the other extensions should then be sub-packages of your single root package. Otherwise it is not possible to link those packages together by using `"directus-extension-pack-price": "1.0.0"`.
Alternatively, you can use local path dependencies as suggested by @dimitrov-adrian.
